### PR TITLE
CI: Simplifying dependencies to speed up conda resolve

### LINKED ISSUE #2670: Speed up dependency resolution
xref #2591

Resolving the Ibis environment is taking around 20 minutes or more. We had problems in the past, particularly with OmniSciDB, and conda performance seems to be getting worse and worse. But still, there shouldn't be any reason to have an insane resolution time like this. Something like 5 seconds is what would make sense.

We discussed in the past about using pip, and we tried unpinning most dependencies (which were pinned to speed up the build), but we didn't have success (see #2591).

Whatever is the solution, we should be able to get an environment that resolves in seconds.

### PR DESCRIPTION
**Pull request by datapythonista**
Closes #2670

Trying to speed up CI builds by speeding up conda solver.

Unpining most dependencies in the base file, and moving few dependencies from the base file to the backend dependencies.

The base file resolves in a bit more than a minute locally.

### CODE REVIEW

**Comment by datapythonista** [comment]
The base environment in Python 3.8 installs in less than 5 minutes with these changes (the resolve, plus downloading and installing). Let's see this exact version on the CI, but seems like in the CI the resolve is way much slower. I don't understand why it should be the case.

**Comment by datapythonista** [comment]
The dependencies in `environment.yml` resolve in around 3 minutes for me locally in Python 3.8. In the CI, in 3.8 seems to be around 20 minutes, and in 3.7, 30 or 40.

I can't really understand how this can be. But personally, I think we should be using pip or mamba in the CI, conda is just making us waste so much of our time with problems and slow builds.

Some examples of the timing:

Impala 3.7: 40 minutes
```
Mon, 08 Mar 2021 16:47:19 GMT
Collecting package metadata (repodata.json): ...working... done
Mon, 08 Mar 2021 17:24:57 GMT
Solving environment: ...working... done
```

pandas 3.8: 20 minutes
```
Mon, 08 Mar 2021 16:43:52 GMT
Collecting package metadata (repodata.json): ...working... done
Mon, 08 Mar 2021 17:03:29 GMT
Solving environment: ...working... done
```

sql 3.7: 30 minutes
```
Mon, 08 Mar 2021 16:45:00 GMT
Collecting package metadata (repodata.json): ...working... done
Mon, 08 Mar 2021 17:44:49 GMT
Solving environment: ...working... done
```

spark 3.8: 20 minutes
```
Mon, 08 Mar 2021 16:43:50 GMT
Collecting package metadata (repodata.json): ...working... done
Mon, 08 Mar 2021 17:02:47 GMT
Solving environment: ...working... done
```

**Comment by datapythonista** [comment]
@jreback, this should fix all problems with the conda solver. tl;dr:
- I removed some dependencies that were not needed
- I removed all the pinning of dependencies added to make the search space smaller (that alone increased the resolve time to up to 40 minutes, with only the required dependencies)
- Instead of installing the base environment first, and then calling conda again for the dependencies of each backend, I dynamically create a single environment file
- I use mamba instead of conda to install the environment file. Mamba is a wrapper around conda that replaces the solver, and uses the same as redhat and other linux distributions instead. This makes the solving take less than 10 seconds instead of +40 minutes

File: ci/deps/pyspark.yml [hunk 1]
```diff
@@ -1 +1 @@
-pyspark>=3.1.1
```
Comment (jreback) [inline ci/deps/pyspark.yml:1]: hmm are we still testing the same as before? cc @icexelloss

**Review by jreback — CHANGES_REQUESTED** [review]
this doesn't affect the min build right?

**Comment by jreback** [comment]
speedup looks great!

File: ci/deps/pyspark.yml (hunk 1 above)
Comment (datapythonista) [reply ci/deps/pyspark.yml:1]: Yes, the pinning is the same for the min, and this was only for speedup. I can leave it if you want, but it's installing the latest version as expected:
```
pyspark                                      3.1.1  pyh44b312d_0
```

**Comment by jreback** [comment]
thanks @datapythonista very nice!

**Pull request merged**